src/ImageRunModal.jsx

```diff
@@ -27,13 +27,112 @@ const units = {
     },
 };
 
+class PublishPortsForm extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            publish: [{ containerPort: null, hostPort: null, protocol: 'TCP' }],
+        };
+        this.removePort = this.removePort.bind(this);
+        this.addPort = this.addPort.bind(this);
+        this.onPortChange = this.onPortChange.bind(this);
+        this.validateAndChange = this.validateAndChange.bind(this);
+    }
+
+    removePort(idx, field, value) {
+        let ports = this.state.publish;
+        ports.splice(idx, 1);
+        if (ports.length === 0) {
+            this.addPort();
+        } else {
+            this.setState({ publish: ports });
+            this.validateAndChange(ports);
+        }
+    }
+
+    addPort() {
+        let ports = this.state.publish;
+        ports.push({ containerPort: null, hostPort: null, protocol: 'TCP' });
+        this.setState({ publish: ports });
+        this.validateAndChange(ports);
+    }
+
+    onPortChange(idx, field, value) {
+        let ports = this.state.publish;
+        ports[idx][field] = value || null;
+        this.setState({ publish: ports });
+        this.validateAndChange(ports);
+    }
+
+    validateAndChange(ports) {
+        if (typeof this.props.onChange !== 'function')
+            return;
+        let validated = ports.filter(port => parseInt(port.containerPort) >= 1 && parseInt(port.containerPort) <= 65535 &&
+                                     parseInt(port.hostPort) >= 1 && parseInt(port.hostPort) <= 65535);
+        this.props.onChange(validated.map(port => port.hostPort + ':' + port.containerPort + '/' + port.protocol));
+    }
+
+    render () {
+        const dialogValues = this.state;
+        return (
+            <div>
```

Comment:
Please make that a `<React.Fragment>`, the `<div>` isn't needed in the final rendered HTML.